velox/exec/tests/HashTableTest.cpp

```diff
@@ -822,3 +826,99 @@ TEST_P(HashTableTest, offsetOverflowLoadTags) {
     insertGroups(*batches.back(), *lookup, *table);
   }
 }
+
+DEBUG_ONLY_TEST_P(HashTableTest, failureInCreateRowPartitions) {
+  // This tests an issue in parallelJoinBuild where an exception in
+  // createRowPartitions could lead to concurrency issues in async table
+  // partitioning threads.
+
+  // It is only relevant when the parallel join build is enabled.
+  if (!GetParam()) {
+    return;
+  }
+
+  // Create a table, and 3 "other" tables.
+  std::unique_ptr<HashTable<false>> topTable;
+  std::vector<std::unique_ptr<BaseHashTable>> otherTables;
+  for (int i = 0; i < 4; i++) {
+    auto batch = vectorMaker_->rowVector(
+        {vectorMaker_->flatVector<int64_t>(10, folly::identity)});
+    std::vector<std::unique_ptr<VectorHasher>> hashers;
+    hashers.push_back(std::make_unique<VectorHasher>(BIGINT(), 0));
+    // Set minTableSizeForParallelJoinBuild to be really small so we can trigger
+    // a parallel join build without needing a lot of data.
+    auto table = HashTable<false>::createForJoin(
+        std::move(hashers), {BIGINT()}, true, false, 1, pool_.get());
+    copyVectorsToTable({batch}, 0, table.get());
+
+    if (!topTable) {
+      topTable = std::move(table);
+    } else {
+      otherTables.emplace_back(std::move(table));
+    }
+  }
+
+  topTable->prepareJoinTable(std::move(otherTables), executor_.get());
+
+  const std::string expectedFailureMessage =
+      "Triggering expected failure in allocation";
+
+  // Fail when allocating memory for the third table.  So we know 2
+  // RowPartitions have been created.
+  std::atomic_int allocateCount{0};
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::common::memory::MemoryPoolImpl::allocateNonContiguous",
+      std::function<void(void*)>(([&](void*) {
+        if (++allocateCount >= 3) {
+          VELOX_FAIL(expectedFailureMessage);
+        }
+      })));
+
+  // Wait until syncWorkItems is called.  This is what waits for any async
+  // threads to complete.
+  std::atomic_bool partitionRowsReady{false};
+  folly::EventCount partitionRowsWait;
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::HashTable::syncWorkItems",
+      std::function<void(void*)>(([&](void*) {
+        partitionRowsReady.store(true);
```

Comment:
You can inject one second delay here to make sure the parallel join build task is running at the background? Thanks!